<h1 align=center><font size = 5>Segmenting and Clustering Neighborhoods in Toronto</font></h1>

## Introduction

This Notebook is to explore, segment, and cluster the neighborhoods in the city of Toronto. For the Toronto neighborhood data, a Wikipedia page, https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M, exists that has all the information we need to explore and cluster the neighborhoods in Toronto.

## Table of Contents

<div class="alert alert-block alert-info" style="margin-top: 20px">

<font size = 3>

1. <a href="#item1">Download and Explore Dataset</a>

2. <a href="#item2">Explore Neighborhoods in New York City</a>

3. <a href="#item3">Analyze Each Neighborhood</a>

4. <a href="#item4">Cluster Neighborhoods</a>

5. <a href="#item5">Examine Clusters</a>    
</font>
</div>

Before we get the data and start exploring it, let's download all the dependencies that we will need.

In [0]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

from bs4 import BeautifulSoup # scraping library

#!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
!pip install geopy
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

#!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

print('Libraries imported.')

Libraries imported.


## 1. Download and Explore Dataset

The Toronto neighborhood data is not readily available on the internet. However
a [Wikipedia page](https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M) exists that has all the information we need to explore and cluster the neighborhoods in Toronto. 

We need to scrape the Wikipedia page and wrangle the data, clean it, and then read it into a pandas dataframe so that it is in a structured format like the New York dataset.

Once the data is in a structured format, you can explore and cluster the neighborhoods in the city of Toronto.

#### Load and explore the data

Let's scrap the Wikipedia's article using BeautifulSoup library.

The Wikipedia's article contains the Toronto neighborhood data which is presented in a Table. We will parse and store the article in a list that contains each row of the table.

In [0]:
# URL to wikipedia page
url='https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'

# Get the entire html of the article as a str
text_result = requests.get(url).text

# Transform the text to html
html_parsed_result = BeautifulSoup(text_result, 'html.parser')

neightborhood_info_table = html_parsed_result.find('table', class_ = 'wikitable')
neightborhood_rows = neightborhood_info_table.find_all('tr')

# Extract the info ('Postcode', 'Borough', 'Neighbourhood') from the table
neightborhood_info = []

for row in neightborhood_rows:
    # Split the row and remove empty str (first and last items)
    info = row.text.split('\n')[1:-1]
    neightborhood_info.append(info)
    
neightborhood_info[0:10]

[['Postcode', 'Borough', 'Neighbourhood'],
 ['M1A', 'Not assigned', 'Not assigned'],
 ['M2A', 'Not assigned', 'Not assigned'],
 ['M3A', 'North York', 'Parkwoods'],
 ['M4A', 'North York', 'Victoria Village'],
 ['M5A', 'Downtown Toronto', 'Harbourfront'],
 ['M6A', 'North York', 'Lawrence Heights'],
 ['M6A', 'North York', 'Lawrence Manor'],
 ['M7A', 'Downtown Toronto', "Queen's Park"],
 ['M8A', 'Not assigned', 'Not assigned']]

#### Tranform the data into a *pandas* dataframe

The next task is essentially transforming this data of nested Python dictionaries into a pandas dataframe.

In [0]:
# Create a Neighborhoods dataframe
neightborhood_info[0][-1] = 'Neighborhood' # change to american spelling
neighborhood_df = pd.DataFrame(neightborhood_info[1:], columns=neightborhood_info[0])

neighborhood_df.head(10)

Postcode           Borough      Neighborhood
0      M1A      Not assigned      Not assigned
1      M2A      Not assigned      Not assigned
2      M3A        North York         Parkwoods
3      M4A        North York  Victoria Village
4      M5A  Downtown Toronto      Harbourfront
5      M6A        North York  Lawrence Heights
6      M6A        North York    Lawrence Manor
7      M7A  Downtown Toronto      Queen's Park
8      M8A      Not assigned      Not assigned
9      M9A         Etobicoke  Islington Avenue

Let's analyze tha data for missing values like 'Not assigned'

In [0]:
not_assigned_boroughs = neighborhood_df.index[neighborhood_df['Borough'] == 'Not assigned']
not_assigned_neighborhoods = neighborhood_df.index[neighborhood_df['Neighborhood'] == 'Not assigned']
not_assigned_neighborhoods_and_borough = not_assigned_boroughs & not_assigned_neighborhoods

print('Shape of the neighborhood DataFrame is {}'.format(neighborhood_df.shape),'\n')
print('Total Postal codes: {}'.format(neighborhood_df['Postcode'].unique().shape[0]))
print('Total Boroughs: {}'.format(neighborhood_df['Borough'].unique().shape[0] - 1)) # substract one because "not assigned" doesn't count
print('Total Neighborhoods: {}'.format(neighborhood_df['Neighborhood'].unique().shape[0] - 1)) # substract one because "not assigned" doesn't count
print('Total rows with Not assigned Borough: {}'.format(not_assigned_boroughs.shape[0]))
print('Total rows with Not assigned Neighborhood: {}'.format(not_assigned_neighborhoods.shape[0]))
print('Total rows with Not assigned Neighborhood and Borough: {}'.format(not_assigned_neighborhoods_and_borough.shape[0]),'\n')

Shape of the neighborhood DataFrame is (287, 3) 

Total Postal codes: 180
Total Boroughs: 10
Total Neighborhoods: 208
Total rows with Not assigned Borough: 77
Total rows with Not assigned Neighborhood: 77
Total rows with Not assigned Neighborhood and Borough: 77 



Now, drop the rows with "Not assigned" Boroughs.

In [0]:
# Drop rows with 'Not assigned'
neighborhood_df.drop(neighborhood_df.index[not_assigned_boroughs], inplace=True)
neighborhood_df.reset_index(drop=True, inplace=True)

neighborhood_df.head(10)

Postcode           Borough      Neighborhood
0      M3A        North York         Parkwoods
1      M4A        North York  Victoria Village
2      M5A  Downtown Toronto      Harbourfront
3      M6A        North York  Lawrence Heights
4      M6A        North York    Lawrence Manor
5      M7A  Downtown Toronto      Queen's Park
6      M9A         Etobicoke  Islington Avenue
7      M1B       Scarborough             Rouge
8      M1B       Scarborough           Malvern
9      M3B        North York   Don Mills North

Replace the "Not assigned" values in the Neighborhood column with corresponding Borough name

In [0]:
# Get indices of 'Not assigned' in the neighborhood column
not_assigned_neighborhoods = neighborhood_df.index[neighborhood_df['Neighborhood'] == 'Not assigned']

# Replace 'Not assigned' with corresponding Borough name
for idx in not_assigned_neighborhoods:
    neighborhood_df['Neighborhood'][idx] = neighborhood_df['Borough'][idx]
    
neighborhood_df.head(10)

Postcode           Borough      Neighborhood
0      M3A        North York         Parkwoods
1      M4A        North York  Victoria Village
2      M5A  Downtown Toronto      Harbourfront
3      M6A        North York  Lawrence Heights
4      M6A        North York    Lawrence Manor
5      M7A  Downtown Toronto      Queen's Park
6      M9A         Etobicoke  Islington Avenue
7      M1B       Scarborough             Rouge
8      M1B       Scarborough           Malvern
9      M3B        North York   Don Mills North

In [0]:
print('Shape of the DataFrame after cleaning is {}'.format(neighborhood_df.shape),'\n')
print('Total Postal codes: {}'.format(neighborhood_df['Postcode'].unique().shape[0]))
print('Total Boroughs: {}'.format(neighborhood_df['Borough'].unique().shape[0]))
print('Total Neighborhoods: {}'.format(neighborhood_df['Neighborhood'].unique().shape[0]))

Shape of the DataFrame after cleaning is (210, 3) 

Total Postal codes: 103
Total Boroughs: 10
Total Neighborhoods: 208


The dataframe has 103 Postal codes but it has 210 rows, because each Postal code can present more than one neighborhood (208 in total). 

Now, let's group the data frame by the Postal code.

In [0]:
# Combining Neighbourhoods based on similar Postcode and Borough
grouped_neighborhood_df = neighborhood_df.groupby(['Postcode', 'Borough'])['Neighborhood'].apply(','.join).reset_index()
grouped_neighborhood_df.columns = ['Postcode','Borough','Neighborhood']
grouped_neighborhood_df.head(10)

Postcode      Borough                                   Neighborhood
0      M1B  Scarborough                                  Rouge,Malvern
1      M1C  Scarborough           Highland Creek,Rouge Hill,Port Union
2      M1E  Scarborough                Guildwood,Morningside,West Hill
3      M1G  Scarborough                                         Woburn
4      M1H  Scarborough                                      Cedarbrae
5      M1J  Scarborough                            Scarborough Village
6      M1K  Scarborough      East Birchmount Park,Ionview,Kennedy Park
7      M1L  Scarborough                  Clairlea,Golden Mile,Oakridge
8      M1M  Scarborough  Cliffcrest,Cliffside,Scarborough Village West
9      M1N  Scarborough                     Birch Cliff,Cliffside West

In [0]:
print('The DataFrame shape is', grouped_neighborhood_df.shape)

The DataFrame shape is (103, 3)


Use geopy library to get the latitude and longitude values of Toronto

In [0]:
# Read Geo-spatial data
coordinates_df = pd.read_csv("http://cocl.us/Geospatial_data")
coordinates_df.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

Adding the coordinates to the neighborhood dataframe, a join is performed using the postcodes as keys

In [0]:
grouped_neighborhood_with_coordinates_df = grouped_neighborhood_df.join(coordinates_df.set_index('Postal Code'), on='Postcode')
grouped_neighborhood_with_coordinates_df.head()

Postcode      Borough                          Neighborhood   Latitude  \
0      M1B  Scarborough                         Rouge,Malvern  43.806686   
1      M1C  Scarborough  Highland Creek,Rouge Hill,Port Union  43.784535   
2      M1E  Scarborough       Guildwood,Morningside,West Hill  43.763573   
3      M1G  Scarborough                                Woburn  43.770992   
4      M1H  Scarborough                             Cedarbrae  43.773136   

   Longitude  
0 -79.194353  
1 -79.160497  
2 -79.188711  
3 -79.216917  
4 -79.239476

#### Get the latitude and longitude values of Toronto using geopy library.

In order to define an instance of the geocoder, we need to define a user_agent. We will name our agent tr_explorer, as shown below.

In [0]:
address = 'Toronto, Canada'

geolocator = Nominatim(user_agent="can_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.653963, -79.387207.


#### Create a map of New York with neighborhoods superimposed on top.

In [0]:
# Create map of Toronto using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# Add markers to map
for lat, lng, borough, neighborhood in zip(
    grouped_neighborhood_with_coordinates_df['Latitude'], 
    grouped_neighborhood_with_coordinates_df['Longitude'], 
    grouped_neighborhood_with_coordinates_df['Borough'], 
    grouped_neighborhood_with_coordinates_df['Neighborhood']
    ):
    
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

## 2. Explore Neighborhoods in Toronto

#### Define Foursquare Credentials and Version

Next, we are going to start utilizing the Foursquare API to explore the neighborhoods and segment them.

In [1]:
# Define Foursquare Credentials and Version
LIMIT = 100

CLIENT_ID = '<CLIENT_ID>'
CLIENT_SECRET = '<CLIENT_SECRET>'
VERSION = '20180605'

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: <CLIENT_ID>
CLIENT_SECRET:<CLIENT_SECRET>


#### Let's create a function to repeat the same process to all the neighborhoods in Toronto

In [0]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # Create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # Make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # Return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

Now write the code to run the above function on each neighborhood and create a new dataframe called toronto_venues.

In [0]:
toronto_venues = getNearbyVenues(names=grouped_neighborhood_with_coordinates_df['Neighborhood'],
                                 latitudes=grouped_neighborhood_with_coordinates_df['Latitude'],
                                 longitudes=grouped_neighborhood_with_coordinates_df['Longitude'],
                                 )

Rouge,Malvern
Highland Creek,Rouge Hill,Port Union
Guildwood,Morningside,West Hill
Woburn
Cedarbrae
Scarborough Village
East Birchmount Park,Ionview,Kennedy Park
Clairlea,Golden Mile,Oakridge
Cliffcrest,Cliffside,Scarborough Village West
Birch Cliff,Cliffside West
Dorset Park,Scarborough Town Centre,Wexford Heights
Maryvale,Wexford
Agincourt
Clarks Corners,Sullivan,Tam O'Shanter
Agincourt North,L'Amoreaux East,Milliken,Steeles East
L'Amoreaux West
Upper Rouge
Hillcrest Village
Fairview,Henry Farm,Oriole
Bayview Village
Silver Hills,York Mills
Newtonbrook,Willowdale
Willowdale South
York Mills West
Willowdale West
Parkwoods
Don Mills North
Flemingdon Park,Don Mills South
Bathurst Manor,Downsview North,Wilson Heights
Northwood Park,York University
CFB Toronto,Downsview East
Downsview West
Downsview Central
Downsview Northwest
Victoria Village
Woodbine Gardens,Parkview Hill
Woodbine Heights
The Beaches
Leaside
Thorncliffe Park
East Toronto
The Danforth West,Riverdale
The Beaches West,Indi

#### Let's check the size of the resulting dataframe

In [0]:
print(toronto_venues.shape)
toronto_venues.head()

(2233, 7)


Neighborhood  Neighborhood Latitude  \
0                         Rouge,Malvern              43.806686   
1  Highland Creek,Rouge Hill,Port Union              43.784535   
2       Guildwood,Morningside,West Hill              43.763573   
3       Guildwood,Morningside,West Hill              43.763573   
4       Guildwood,Morningside,West Hill              43.763573   

   Neighborhood Longitude                  Venue  Venue Latitude  \
0              -79.194353                Wendy's       43.807448   
1              -79.160497  Royal Canadian Legion       43.782533   
2              -79.188711      G & G Electronics       43.765309   
3              -79.188711             Marina Spa       43.766000   
4              -79.188711       Big Bite Burrito       43.766299   

   Venue Longitude        Venue Category  
0       -79.199056  Fast Food Restaurant  
1       -79.163085                   Bar  
2       -79.191537     Electronics Store  
3       -79.191000                   Spa  
4       -79.190720    Mexican Restaurant

Let's check how many venues were returned for each neighborhood

In [0]:
toronto_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Adelaide,King,Richmond                                                100   
Agincourt                                                               4   
Agincourt North,L'Amoreaux East,Milliken,Steele...                      2   
Albion Gardens,Beaumond Heights,Humbergate,Jame...                     10   
Alderwood,Long Branch                                                   9   
Bathurst Manor,Downsview North,Wilson Heights                          21   
Bayview Village                                                         4   
Bedford Park,Lawrence Manor East                                       24   
Berczy Park                                                            55   
Birch Cliff,Cliffside West                                              4   
Bloordale Gardens,Eringate,Markland Wood,Old Bu...                      9   
Brockton,Exhibition Place,Parkdale Village                             25   
Business Reply Mail Processing Centre 969 Eastern                      17   
CFB Toronto,Downsview East                                              3   
CN Tower,Bathurst Quay,Island airport,Harbourfr...                     17   
Cabbagetown,St. James Town                                             45   
Caledonia-Fairbanks                                                     4   
Canada Post Gateway Processing Centre                                  12   
Cedarbrae                                                               8   
Central Bay Street                                                     80   
Chinatown,Grange Park,Kensington Market                                88   
Christie                                                               18   
Church and Wellesley                                                   86   
Clairlea,Golden Mile,Oakridge                                          10   
Clarks Corners,Sullivan,Tam O'Shanter                                  12   
Cliffcrest,Cliffside,Scarborough Village West                           3   
Cloverdale,Islington,Martin Grove,Princess Gard...                      1   
Commerce Court,Victoria Hotel                                         100   
Davisville                                                             35   
Davisville North                                                        9   
Deer Park,Forest Hill SE,Rathnelly,South Hill,S...                     14   
Del Ray,Keelesdale,Mount Dennis,Silverthorn                             3   
Design Exchange,Toronto Dominion Centre                               100   
Don Mills North                                                         6   
Dorset Park,Scarborough Town Centre,Wexford Hei...                      5   
Dovercourt Village,Dufferin                                            18   
Downsview Central                                                       2   
Downsview Northwest                                                     5   
Downsview West                                                          6   
Downsview,North Park,Upwood Park                                        4   
East Birchmount Park,Ionview,Kennedy Park                               6   
East Toronto                                                            4   
Emery,Humberlea                                                         1   
Fairview,Henry Farm,Oriole                                             65   
First Canadian Place,Underground city                                 100   
Flemingdon Park,Don Mills South                                        22   
Forest Hill North,Forest Hill West                                      4   
Glencairn                                                               5   
Guildwood,Morningside,West Hill                                         8   
Harbord,University of Toronto                                          36   
Harbourfront                                    

#### Let's find out how many unique categories can be curated from all the returned venues

In [0]:
print('There are {} uniques categories.'.format(len(toronto_venues['Venue Category'].unique())))

There are 265 uniques categories.


## 3. Analyze Each Neighborhood

In [0]:
# One hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# Add neighborhood column back to dataframe
toronto_onehot['Neighborhood'] = toronto_venues['Neighborhood'] 

# Move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot.head()

Yoga Studio  Accessories Store  Afghan Restaurant  Airport  \
0            0                  0                  0        0   
1            0                  0                  0        0   
2            0                  0                  0        0   
3            0                  0                  0        0   
4            0                  0                  0        0   

   Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0                   0             0               0                0   
1                   0             0               0                0   
2                   0             0               0                0   
3                   0             0               0                0   
4                   0             0               0                0   

   Airport Terminal  American Restaurant  Antique Shop  Aquarium  Art Gallery  \
0                 0                    0             0         0            0   
1                 0                    0             0         0            0   
2                 0                    0             0         0            0   
3                 0                    0             0         0            0   
4                 0                    0             0         0            0   

   Arts & Crafts Store  Asian Restaurant  Athletics & Sports  Auto Workshop  \
0                    0                 0                   0              0   
1                    0                 0                   0              0   
2                    0                 0                   0              0   
3                    0                 0                   0              0   
4                    0                 0                   0              0   

   BBQ Joint  Baby Store  Bagel Shop  Bakery  Bank  Bar  Baseball Field  \
0          0           0           0       0     0    0               0   
1          0           0           0       0     0    1               0   
2          0           0           0       0     0    0               0   
3          0           0           0       0     0    0               0   
4          0           0           0       0     0    0               0   

   Baseball Stadium  Basketball Court  Basketball Stadium  Beach  \
0                 0                 0                   0      0   
1                 0                 0                   0      0   
2                 0                 0                   0      0   
3                 0                 0                   0      0   
4                 0                 0                   0      0   

   Bed & Breakfast  Beer Bar  Beer Store  Belgian Restaurant  Bike Shop  \
0                0         0           0                   0          0   
1                0         0           0                   0          0   
2                0         0           0                   0          0   
3                0         0           0                   0          0   
4                0         0           0                   0          0   

   Bistro  Boat or Ferry  Bookstore  Boutique  Brazilian Restaurant  \
0       0              0          0         0                     0   
1       0              0          0         0                     0   
2       0              0          0         0                     0   
3       0              0          0         0                     0   
4       0              0          0         0                     0   

   Breakfast Spot  Brewery  Bridal Shop  Bubble Tea Shop  Building  \
0               0        0            0                0         0   
1               0        0            0                0         0   
2               0        0            0                0         0   
3               0        0            0                0         0   
4               0        0            0                0         0   

   Burger Joint  Burrito Place  Bus Line  Bus Station  Bus

And let's examine the new dataframe size.

In [0]:
toronto_onehot.shape

(2233, 265)

#### Next, let's group rows by neighborhood and by taking the mean of the frequency of occurrence of each category

In [0]:
toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()
toronto_grouped

Neighborhood  Yoga Studio  \
0                              Adelaide,King,Richmond     0.000000   
1                                           Agincourt     0.000000   
2   Agincourt North,L'Amoreaux East,Milliken,Steel...     0.000000   
3   Albion Gardens,Beaumond Heights,Humbergate,Jam...     0.000000   
4                               Alderwood,Long Branch     0.000000   
5       Bathurst Manor,Downsview North,Wilson Heights     0.000000   
6                                     Bayview Village     0.000000   
7                    Bedford Park,Lawrence Manor East     0.000000   
8                                         Berczy Park     0.000000   
9                          Birch Cliff,Cliffside West     0.000000   
10  Bloordale Gardens,Eringate,Markland Wood,Old B...     0.000000   
11         Brockton,Exhibition Place,Parkdale Village     0.040000   
12  Business Reply Mail Processing Centre 969 Eastern     0.058824   
13                         CFB Toronto,Downsview East     0.000000   
14  CN Tower,Bathurst Quay,Island airport,Harbourf...     0.000000   
15                         Cabbagetown,St. James Town     0.000000   
16                                Caledonia-Fairbanks     0.000000   
17              Canada Post Gateway Processing Centre     0.000000   
18                                          Cedarbrae     0.000000   
19                                 Central Bay Street     0.012500   
20            Chinatown,Grange Park,Kensington Market     0.000000   
21                                           Christie     0.000000   
22                               Church and Wellesley     0.011628   
23                      Clairlea,Golden Mile,Oakridge     0.000000   
24              Clarks Corners,Sullivan,Tam O'Shanter     0.000000   
25      Cliffcrest,Cliffside,Scarborough Village West     0.000000   
26  Cloverdale,Islington,Martin Grove,Princess Gar...     0.000000   
27                      Commerce Court,Victoria Hotel     0.000000   
28                                         Davisville     0.000000   
29                                   Davisville North     0.000000   
30  Deer Park,Forest Hill SE,Rathnelly,South Hill,...     0.000000   
31        Del Ray,Keelesdale,Mount Dennis,Silverthorn     0.000000   
32            Design Exchange,Toronto Dominion Centre     0.000000   
33                                    Don Mills North     0.000000   
34  Dorset Park,Scarborough Town Centre,Wexford He...     0.000000   
35                        Dovercourt Village,Dufferin     0.000000   
36                                  Downsview Central     0.000000   
37                                Downsview Northwest     0.000000   
38                                     Downsview West     0.000000   
39                   Downsview,North Park,Upwood Park     0.000000   
40          East Birchmount Park,Ionview,Kennedy Park     0.000000   
41                                       East Toronto     0.000000   
42                                    Emery,Humberlea     0.000000   
43                         Fairview,Henry Farm,Oriole     0.000000   
44              First Canadian Place,Underground city     0.000000   
45                    Flemingdon Park,Don Mills South     0.000000   
46                 Forest Hill North,Forest Hill West     0.000000   
47                                          Glencairn     0.000000   
48                    Guildwood,Morningside,West Hill     0.000000   
49                      Harbord,University of Toronto     0.027778   
50                                       Harbourfront     0.000000   
51    Harbourfront East,Toronto Islands,Union Station     0.000000   
52                       High Park,The Junction South     0.000000   
53               Highland Creek,Rouge Hill,Port Union     0.000000   
54                                  Hillcrest Village     0.000000   
55         Humber Bay Shores,Mimico South,New Toronto     0.000000   
56  Humber Bay,King's Mill Park,Kingsway Park Sout.

#### Let's confirm the new size

In [0]:
toronto_grouped.shape

(100, 265)

#### Let's print each neighborhood along with the top 5 most common venues

In [0]:
num_top_venues = 5

for hood in toronto_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = toronto_grouped[toronto_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Adelaide,King,Richmond----
             venue  freq
0      Coffee Shop  0.07
1       Restaurant  0.05
2  Thai Restaurant  0.04
3             Café  0.04
4              Bar  0.03


----Agincourt----
                       venue  freq
0  Latin American Restaurant  0.25
1                     Lounge  0.25
2             Breakfast Spot  0.25
3               Skating Rink  0.25
4                Yoga Studio  0.00


----Agincourt North,L'Amoreaux East,Milliken,Steeles East----
                       venue  freq
0                       Park   0.5
1                 Playground   0.5
2             Medical Center   0.0
3         Miscellaneous Shop   0.0
4  Middle Eastern Restaurant   0.0


----Albion Gardens,Beaumond Heights,Humbergate,Jamestown,Mount Olive,Silverstone,South Steeles,Thistletown----
                  venue  freq
0         Grocery Store   0.2
1   Japanese Restaurant   0.1
2              Pharmacy   0.1
3           Pizza Place   0.1
4  Fast Food Restaurant   0.1


----Alderwood,Long B

#### Let's put that into a *pandas* dataframe

First, let's write a function to sort the venues in descending order.

In [0]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

Now let's create the new dataframe and display the top 10 venues for each neighborhood.

In [0]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood  \
0                             Adelaide,King,Richmond   
1                                          Agincourt   
2  Agincourt North,L'Amoreaux East,Milliken,Steel...   
3  Albion Gardens,Beaumond Heights,Humbergate,Jam...   
4                              Alderwood,Long Branch   

       1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
0                Coffee Shop            Restaurant                  Café   
1  Latin American Restaurant          Skating Rink        Breakfast Spot   
2                       Park            Playground   Distribution Center   
3              Grocery Store   Japanese Restaurant  Fast Food Restaurant   
4                Pizza Place           Coffee Shop          Skating Rink   

  4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
0       Thai Restaurant                   Bar            Steakhouse   
1                Lounge         Deli / Bodega   Dumpling Restaurant   
2         Deli / Bodega      Department Store          Dessert Shop   
3              Pharmacy           Pizza Place        Sandwich Place   
4                   Pub                   Gym    Athletics & Sports   

  7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
0      Sushi Restaurant           Pizza Place                Lounge   
1             Drugstore            Donut Shop      Doner Restaurant   
2    Dim Sum Restaurant                 Diner        Discount Store   
3            Beer Store   Fried Chicken Joint        Discount Store   
4              Pharmacy        Sandwich Place         Women's Store   

  10th Most Common Venue  
0     Seafood Restaurant  
1                Dog Run  
2                Dog Run  
3       Department Store  
4     Dim Sum Restaurant

## 4. Cluster Neighborhoods


Run k-means to cluster the neighborhood into 5 clusters.

In [0]:
# Set number of clusters
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighborhood', 1)

# Run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# Check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10]

array([2, 2, 0, 2, 2, 2, 2, 2, 2, 2], dtype=int32)

Let's create a new dataframe that includes the cluster as well as the top 10 venues for each neighborhood.

In [0]:
# Add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged = grouped_neighborhood_with_coordinates_df

# Merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

toronto_merged.head() # check the last columns!

Postcode      Borough                          Neighborhood   Latitude  \
0      M1B  Scarborough                         Rouge,Malvern  43.806686   
1      M1C  Scarborough  Highland Creek,Rouge Hill,Port Union  43.784535   
2      M1E  Scarborough       Guildwood,Morningside,West Hill  43.763573   
3      M1G  Scarborough                                Woburn  43.770992   
4      M1H  Scarborough                             Cedarbrae  43.773136   

   Longitude  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0 -79.194353             1.0  Fast Food Restaurant         Women's Store   
1 -79.160497             2.0                   Bar         Women's Store   
2 -79.188711             2.0    Mexican Restaurant   Rental Car Location   
3 -79.216917             2.0           Coffee Shop     Korean Restaurant   
4 -79.239476             2.0                  Bank   Fried Chicken Joint   

  3rd Most Common Venue        4th Most Common Venue  \
0          Dance Studio  Eastern European Restaurant   
1         Deli / Bodega            Electronics Store   
2                   Spa               Breakfast Spot   
3               Dog Run             Department Store   
4    Athletics & Sports             Hakka Restaurant   

         5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0          Dumpling Restaurant             Drugstore            Donut Shop   
1  Eastern European Restaurant   Dumpling Restaurant             Drugstore   
2                 Intersection        Medical Center                  Bank   
3                 Dessert Shop    Dim Sum Restaurant                 Diner   
4         Caribbean Restaurant       Thai Restaurant           Gas Station   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0      Doner Restaurant               Dog Run    Distribution Center  
1            Donut Shop      Doner Restaurant                Dog Run  
2     Electronics Store     Convenience Store         Cosmetics Shop  
3        Discount Store   Distribution Center       Doner Restaurant  
4                Bakery    Dim Sum Restaurant          Deli / Bodega

Finally, let's visualize the resulting clusters

In [0]:
# Create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# Set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# Add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighborhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        fill=True,
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

## 5. Examine Clusters

Now, we can examine each cluster and determine the discriminating venue categories that distinguish each cluster. Based on the defining categories, we can then assign a name to each cluster.

Cluster 1

In [0]:

toronto_merged.loc[toronto_merged['Cluster Labels'] == 0, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
14       Scarborough             0.0                  Park   
23        North York             0.0                  Park   
25        North York             0.0              Bus Stop   
30        North York             0.0                  Park   
37      East Toronto             0.0                  Park   
40         East York             0.0                  Park   
44   Central Toronto             0.0                  Park   
48   Central Toronto             0.0                  Park   
50  Downtown Toronto             0.0                  Park   
72        North York             0.0                  Park   
74              York             0.0                  Park   
79        North York             0.0      Basketball Court   
90         Etobicoke             0.0                  Park   

   2nd Most Common Venue 3rd Most Common Venue        4th Most Common Venue  \
14            Playground   Distribution Center                Deli / Bodega   
23     Convenience Store     Electronics Store                         Bank   
25     Food & Drink Shop                  Park          Dumpling Restaurant   
30               Airport           Snack Place          Dumpling Restaurant   
37                 Trail                   Pub            Health Food Store   
40     Convenience Store           Coffee Shop                Women's Store   
44           Swim School              Bus Line                        Diner   
48                 Trail            Playground                  Curling Ice   
50                 Trail            Playground                  Curling Ice   
72                Bakery   Japanese Restaurant                   Playground   
74         Women's Store                Market  Eastern European Restaurant   
79                  Park                Bakery   Construction & Landscaping   
90            Smoke Shop                 River                    Drugstore   

          5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
14             Department Store          Dessert Shop    Dim Sum Restaurant   
23                 Dessert Shop    Dim Sum Restaurant                 Diner   
25                    Drugstore            Donut Shop      Doner Restaurant   
30  Eastern European Restaurant             Drugstore            Donut Shop   
37                 Dance Studio         Deli / Bodega      Department Store   
40             Department Store          Dessert Shop    Dim Sum Restaurant   
44                Deli / Bodega      Department Store          Dessert Shop   
48                 Dance Studio         Deli / Bodega      Department Store   
50                 Dance Studio         Deli / Bodega      Department Store   
72                          Pub        Discount Store      Department Store   
74          Dumpling Restaurant             Drugstore            Donut Shop   
79                Women's Store          Dessert Shop    Dim Sum Restaurant   
90                   Donut Shop      Doner Restaurant               Dog Run   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
14                 Diner        Discount Store                Dog Run  
23        Discount Store   Distribution Center                Dog Run  
25               Dog Run     Electronics Store         Discount Store  
30      Doner Restaurant               Dog Run            Curling Ice  
37          Dessert Shop    Dim Sum Restaurant                  Diner  
40                 Diner        Discount Store    Distribution Center  
44    Dim Sum Restaurant        Discount Store      Electronics Store  
48          Dessert Shop    Dim Sum Restaurant                  Diner  
50          Dessert Shop    Dim Sum Restaurant                  Diner  
72          Dessert Shop    Dim Sum Restaurant                  Diner  
74      Doner Restaurant               Dog Run           Dance Studio  
79                 Diner        Discount Store    Distribution Cen

Cluster 2

In [0]:

toronto_merged.loc[toronto_merged['Cluster Labels'] == 1, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0  Scarborough             1.0  Fast Food Restaurant         Women's Store   

  3rd Most Common Venue        4th Most Common Venue 5th Most Common Venue  \
0          Dance Studio  Eastern European Restaurant   Dumpling Restaurant   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0             Drugstore            Donut Shop      Doner Restaurant   

  9th Most Common Venue 10th Most Common Venue  
0               Dog Run    Distribution Center

Cluster 3

In [0]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 2, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels      1st Most Common Venue  \
1         Scarborough             2.0                        Bar   
2         Scarborough             2.0         Mexican Restaurant   
3         Scarborough             2.0                Coffee Shop   
4         Scarborough             2.0                       Bank   
5         Scarborough             2.0              Grocery Store   
6         Scarborough             2.0          Convenience Store   
7         Scarborough             2.0                     Bakery   
8         Scarborough             2.0                      Motel   
9         Scarborough             2.0                       Café   
10        Scarborough             2.0          Indian Restaurant   
11        Scarborough             2.0  Middle Eastern Restaurant   
12        Scarborough             2.0  Latin American Restaurant   
13        Scarborough             2.0                Pizza Place   
15        Scarborough             2.0       Fast Food Restaurant   
17         North York             2.0                Golf Course   
18         North York             2.0             Clothing Store   
19         North York             2.0         Chinese Restaurant   
20         North York             2.0                  Cafeteria   
22         North York             2.0           Ramen Restaurant   
24         North York             2.0              Grocery Store   
26         North York             2.0                       Café   
27         North York             2.0                Coffee Shop   
28         North York             2.0                       Bank   
29         North York             2.0             Massage Studio   
31         North York             2.0              Grocery Store   
33         North York             2.0              Grocery Store   
34         North York             2.0               Intersection   
35          East York             2.0                Pizza Place   
36          East York             2.0                       Park   
38          East York             2.0        Sporting Goods Shop   
39          East York             2.0          Indian Restaurant   
41       East Toronto             2.0           Greek Restaurant   
42       East Toronto             2.0                       Park   
43       East Toronto             2.0                       Café   
45    Central Toronto             2.0                       Park   
46    Central Toronto             2.0             Clothing Store   
47    Central Toronto             2.0                Pizza Place   
49    Central Toronto             2.0                        Pub   
51   Downtown Toronto             2.0                Coffee Shop   
52   Downtown Toronto             2.0                Coffee Shop   
53   Downtown Toronto             2.0                Coffee Shop   
54   Downtown Toronto             2.0                Coffee Shop   
55   Downtown Toronto             2.0                Coffee Shop   
56   Downtown Toronto             2.0                Coffee Shop   
57   Downtown Toronto             2.0                Coffee Shop   
58   Downtown Toronto             2.0                Coffee Shop   
59   Downtown Toronto             2.0                Coffee Shop   
60   Downtown Toronto             2.0                Coffee Shop   
61   Downtown Toronto             2.0                Coffee Shop   
62         North York             2.0                Coffee Shop   
64    Central Toronto             2.0              Jewelry Store   
65    Central Toronto             2.0                       Café   
66   Downtown Toronto             2.0                       Café   
67   Downtown Toronto             2.0                        Bar   
68   Downtown Toronto             2.0            Airport Service   
69   Downtown Toronto             2.0                Coffee Shop   
70   Downtown Toronto             2.0                Coffee Shop   
71         North York             2.0             Clothing Store   
73

Cluster 4

In [0]:

toronto_merged.loc[toronto_merged['Cluster Labels'] == 3, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
32  North York             3.0            Food Truck        Baseball Field   
97  North York             3.0        Baseball Field         Women's Store   

   3rd Most Common Venue 4th Most Common Venue        5th Most Common Venue  \
32         Women's Store   Distribution Center                 Dessert Shop   
97         Deli / Bodega     Electronics Store  Eastern European Restaurant   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
32    Dim Sum Restaurant                 Diner        Discount Store   
97   Dumpling Restaurant             Drugstore            Donut Shop   

   9th Most Common Venue 10th Most Common Venue  
32               Dog Run   Ethiopian Restaurant  
97      Doner Restaurant                Dog Run

Cluster 5

In [0]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 4, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
63  Central Toronto             4.0                Garden   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
63         Women's Store        Discount Store         Deli / Bodega   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
63      Department Store          Dessert Shop    Dim Sum Restaurant   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
63                 Diner   Distribution Center            Curling Ice

Clusters 2 and 0 are the biggest, cluster 3 has only two neighbourhoods and
clusters 4 and 1 have just one neighbourhood each